src/useParam.ts

```diff
@@ -0,0 +1,37 @@
+import deepEqual from 'deep-equal'
+import { useHistory, useLocation } from 'react-router-dom'
+import { UrlParam } from './params'
+
+type SetValueFn<T> = (value: T, method?: 'push' | 'replace') => void
+
+const useParam = <T>(urlParam: UrlParam<T>): [T, SetValueFn<T>] => {
+  const history = useHistory()
+  const location = useLocation()
+  const params = new URLSearchParams(location.search)
+  const rawValue = params.get(urlParam.name)
+
+  const setValue: SetValueFn<T> = (value, method = 'push') => {
+    const encodedValue = urlParam.encode(value)
+
+    // Check if we don't want to show the default value in the URL
+    const newValue =
+      !urlParam.showDefaultValueInQuery && deepEqual(urlParam.defaultValue, value)
```

Comment:
In what scenario would we want to use `showDefaultValueInQuery`? I would leave it our for now until we actually run into a scenario where we would need it.

Reply:
Indeed we don't need it for now, but when we publish this as a package, i'd be sure this is a needed feature. So I'll leave it in for now